totoify_grafeas/totoifylib.py

```diff
@@ -0,0 +1,266 @@
+
+<Started>
+  Sep 5, 2020
+
+<Copyright>
+  See LICENSE for licensing information.
+
+<Purpose>
+  Conversion library for Grafeas occurrences and and in-toto link metadata.
+"""
+
+from in_toto.models.link import Link
+from in_toto.models.metadata import Metablock
+import requests
+import json
+
+
+class GrafeasLink:
+  """Wrapper for in-toto 'signed' fields.
+
+  Used in the GrafeasInTotoOccurrence class to organize in-toto "signed"
+  attributes (corresponding to in_toto.models.link.Link).
+
+  Attributes:
+    command: A list of arguments used in the link's command.
+    environment: A dict of the environment where the command was executed.
+    byproducts: A dict containing byproducts of the command.
+    materials: A list of materials used in the command.
+    products: A list of dicts containing resource URI and hashes.
+  """
+
+  materials = None
+  products = None
+  command = None
+  byproducts = None
+  environment = None
+
+  def __init__(self, materials=None, products=None, command=None,
+               byproducts=None, environment=None):
+    """Takes in materials, products, command, byproducts and initalizes
+    to the GrafeasLink fields."""
+
+    if materials is None:
+      self.materials = []
+    elif isinstance(materials, list):
+      self.materials = materials
+
+    if products is None:
+      self.products = []
+    elif isinstance(products, list):
+      self.products = products
+
+    if command is None:
+      self.command = []
+    elif isinstance(command, list):
+      self.command = command
+
+    if byproducts is None:
+      self.byproducts = {"custom_values": {}}
+    elif isinstance(byproducts, dict):
+      self.byproducts = byproducts
+
+    if environment is None:
+      self.environment = {"custom_values": {}}
+    elif isinstance(environment, dict):
+      self.environment = environment
+
+  @staticmethod
+  def from_link(link):
+    link.validate()
+
+    grafeas_link = GrafeasLink()
+
+    # materials and products
+    for item in link.materials:
+      grafeas_link.materials.append({"resource_uri": item,
+                                     "hashes": link.materials[item]})
+
+    for item in link.products:
+      grafeas_link.products.append({"resource_uri": item,
+                                    "hashes": link.products[item]})
+
+    grafeas_link.command = link.command
+
+    for key, value in link.byproducts.items():
+      if key == "return-value":
+        grafeas_link.byproducts["custom_values"][key] = str(value)
+      else:
+        grafeas_link.byproducts["custom_values"][key] = value
+
+    for key, value in link.environment.items():
+      grafeas_link.environment["custom_values"][key] = value
+
+    return grafeas_link
+
+  def __repr__(self):
+    return json.dumps({
+        "command": self.command,
+        "materials": self.materials,
+        "products": self.products,
+        "environment": self.environment,
+        "byproducts": self.byproducts
+    })
```

Comment:
`GrafeasInTotoOccurrence.to_json` returns JSON while this creates a dictionary... I'm worried about confusion in what's returned, there.

Reply:
I've added to_dict to GrafeasInTotoOccurrence to avoid confusion and this is now used within to_json. This is likely a good thing regardless, to have access to that representation.